index.d.ts

```diff
@@ -1,12 +1,49 @@
-import { Parser } from 'acorn' 
+import * as acorn from 'acorn';
 
-declare const jsx: (options?: jsx.Options) => (BaseParser: typeof Parser) => typeof Parser;
+export const jsx: (options?: jsx.Options) => (BaseParser: typeof acorn.Parser) => typeof AcornJsxParser
 
-declare namespace jsx {
-  interface Options {
-    allowNamespacedObjects?: boolean;
-    allowNamespaces?: boolean;
-  }
+export interface Options {
+  allowNamespacedObjects?: boolean;
+  allowNamespaces?: boolean;
 }
 
-export = jsx;
+export const tokTypes: {
+  jsxName: acorn.TokenType,
+  jsxText: acorn.TokenType,
+  jsxTagEnd: acorn.TokenType,
+  jsxTagStart: acorn.TokenType
+} & typeof acorn.tokTypes;
+
+export const tokContexts: {
+  tc_oTag: acorn.TokContext,
+  tc_cTag: acorn.TokContext,
+  tc_expr: acorn.TokContext
+};
+
+export class AcornJsxParser extends acorn.Parser {
+  static readonly acornJsx: {
+    tokTypes: typeof tokTypes;
+    tokContexts: typeof tokContexts
+  };
+
+  jsx_readToken(): string;
+  jsx_readNewLine(normalizeCRLF: boolean): void;
+  jsx_readString(quote: number): void;
+  jsx_readEntity(): string;
+  jsx_readWord(): void;
+  jsx_parseIdentifier(): acorn.Node;
+  jsx_parseNamespacedName(): acorn.Node;
+  jsx_parseElementName(): acorn.Node | string;
+  jsx_parseAttributeValue(): acorn.Node;
+  jsx_parseEmptyExpression(): acorn.Node;
+  jsx_parseExpressionContainer(): acorn.Node;
+  jsx_parseAttribute(): acorn.Node;
+  jsx_parseOpeningElementAt(startPos: number, startLoc?: acorn.SourceLocation): acorn.Node;
+  jsx_parseClosingElementAt(startPos: number, startLoc?: acorn.SourceLocation): acorn.Node;
+  jsx_parseElementAt(startPos: number, startLoc?: acorn.SourceLocation): acorn.Node;
+  jsx_parseText(): acorn.Node;
+  jsx_parseElement(): acorn.Node;
+}
+
+export as namespace jsx;
+export default jsx;
```

Comment:
This declares a default export, but this package doesn’t have a default export, it only has `module.exports`, which should be declared using `exports = jsx`

Reply:
Thanks so much for replying on this, @remcohaszing .

The fixes you provided all sound reasonable, but I am tripped up with the namespace issue in particular, so I would be grateful for any help toward my better understanding why my approach isn't working or if there may indeed be some required export style missing for ESM support.

When targeting acorn-jsx using the exporting approach you mention, and importing via `import jsx from 'acorn-jsx';` (the default namespace approach being that called for with CJS `module.exports` per https://nodejs.org/api/esm.html#commonjs-namespaces ), I get this error:

> This module is declared with using 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag.

I can only imagine something is missing, at least for an ESM context, that such an error will be reported and that some non-default import style should be recommended. And if I switch to the `import * as jsx from 'acorn-jsx'` style, the code, as would be expected, fails.

Reply:
I see. The whole JavaScript module system is a bit of a mess right now. 🥲

This project exposes CommonJS.

Without TypeScript, the correct way to use this in a CommonJS code base is:

```js
// my-file.cjs
const jsx = require('acorn-jsx')
```

Because Node.js allows you to import CommonJS `module.exports` as a default import, the way to use it with ESM in Node.js is:

```js
// my-file.mjs
import jsx from 'acorn-jsx'
```

Beware alternative environments, such as module bundlers, have their own ways of dealing with interoperability.

Now let’s say we’re writing a TypeScript project that compiles to CommonJS. `tsconfig.json` should contain the following content:

```jsonc
{
  "compilerOptions": {
    "module": "CommonJS",
    "moduleResolution": "Node",
  }
}
```

In your code, you can write the following:

```ts
import jsx = require('acorn-jsx')
```

I think nobody really likes this syntax though. As far as I can remember I’ve only seen this syntax in type tests on DefinitelyTyped.

Because a namespace is defined, it’s also possible to use `acorn-jsx` using the following syntax:

```ts
import * as jsx from 'acorn-jsx'
import { 
  // This is an actual value
  tokTypes,
  // This is a type only
  TokContexts,
} from 'acorn-jsx'

// If tokTypes was a function call, it would be bound.
console.log(tokTypes)
```

This roughly compiles to:

```js
const jsx_1 = require('acorn-jsx')
const jsx_2 = require('acorn-jsx')

// If tokTypes was a function call, it would be bound.
console.log(jsx_2.tokTypes)
```

Notice this syntax is different from how Node.js native ESM can import CJS. It uses a star import, not a default import.

Some people prefer to use a default import, mostly because it looks nicer. (We’re still talking about CJS output here)

```jsonc
{
  "compilerOptions": {
    "esModuleInterop": true,
    "module": "CommonJS",
    "moduleResolution": "Node",
  }
}
```

This means the import can now be written as this:

```ts
import jsx from 'acorn-jsx'

console.log(jsx)
console.log(jsx.tokTypes)
```

This roughly compiles to:

```js
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

const jsx_1 = _interopRequireDefault(require('acorn-jsx'))

console.log(jsx_1.default)
console.log(jsx_1.default.tokTypes)
```

So basically, at this point `import jsx from 'acorn-jsx'` is equal to `import * as jsx from 'acorn-jsx'`.

But let’s say you really want to use ESM the way Node.js supports it. Your `package.json` contains the following:

```json
{
  "type": "module"
}
```

Your `tsconfig.json` should contain the following:

```jsonc
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "module": "Node16", // Or "NodeNext" or "ESNext" or "None"
    "moduleResolution": "Node",
  }
}
```

The `allowSyntheticDefaultImports` option specifies that even if a package export is typed as CommonJS using `export =`, it can be imported using a default import syntax.

This means you can now write the following:

```ts
import jsx from 'acorn-jsx'

console.log(jsx.tokTypes)
```

The output will be compatible with Node.js ESM:

```js
import jsx from 'acorn-jsx'

console.log(jsx.tokTypes)
```

I hope this little overview will help you not only understand why the namespace was used like this, but also when to choose what compiler options. :)

Reply:
Ok, makes sense. (And thanks for the helpful explanation; the extra reinforcement is much appreciated!) I forget that the declaration file is not always being compiled from a source (as in this project), so we may have to indicate what is actually there as far as exports. :-)  Still, I would have thought `module.exports`-style would be treated by TypeScript as in Node--i.e., as a default (too). At least I would have thought it wouldn't create problems to make the file work for both ESM and CJS since importing ESM code will know how to handle `module.exports` (and indeed I see no errors when doing so) and I don't see how CJS would treat it differently. I'd be curious to know what unintended consequence might happen if specifying there was a default here. Is it just a matter of accurate semantics?

In our case we're using just JavaScript (ESM) + JSDoc, so we have to use the default import syntax. I guess the eventual way around needing special config (to allow synthetic default imports) is if this package adds an ESM export (and corresponding declaration file).

Reply:
Adding types for a default export indicates the package should be consumed like this in CJS:

```js
const jsx = require('estree-jsx').default
```

This is both incorrect and will break existing packages.

It will complain about the following correct TypeScript code:

```ts
import jsx = require('estree-jsx')
```

and when `checkJs` is enabled, it will also complain about the following JavaScript code:

```ts
const jsx = require('estree-jsx')
```

The `exports = jsx` syntax is intended for typing values assigned to `module.exports`. For ESM packages consuming CJS, the `allowSyntheticDefaultImports` exists. This is not a workaround for consuming this specific package, it’s the way in TypeScript to deal with any package that assigns a value to `module.exports`.

You could try using `lodash` in your package for example. It works the same way (assigns to `module.exports`, needs to be imported as default in ESM, and is typed using `export = _`.

Reply:
I see, thank you very much. It's finally starting to come together. So although, as per the docs, `allowSyntheticDefaultImports` doesn't affect builds and is only for types, the exports do affect builds. This wasn't obvious to me as I thought any compilation by TypeScript (which is not always needed, such as in our case of working with plain JavaScript) might just look at the source itself. I guess this is a bit like Rollup where you sometimes need to add your own config to determine how CJS will be handled, even though in an ideal world, it would mostly be handled by the compiler.

Reply:
Let’s just hope that within a year or so most of the JavaScript ecosystem has migrated to using ESM, so we can forget about CJS compatibility :sweat_smile: